README.rst
A statistic package for python with enphasis on timeseries analysis.
Built around numpy_, it provides several back-end timeseries classes including R-based objects via rpy2_.
It is shipped with a domain specific language for timeseries analysis
and manipulation built on to of ply_.
It requires Python 2.6 and up, including Python 3 versions.

--

:Documentation: http://packages.python.org/dynts/
:Dowloads: http://pypi.python.org/pypi/dynts/
:Source: http://github.com/quantmind/dynts
:Keywords: timeseries, quantitative, finance, statistics, numpy, R, web

--


.. contents::
    :local:


Timeseries
========================

To create a timeseries object directly::

	>>> from dynts import timeseries
	>>> ts = timeseries('test')
	>>> ts.type
	'numpy'
	>>> ts.name
	'test'
	>>> ts
	TimeSeries:numpy:test
	>>> str(ts)
	'test'


DSL
=======
The package comes with a Domain-Specific-Language (DSL_) dedicated
to timeserie analysis and manipulation.
This is a simple multiplication::
	
	>>> import dynts
	>>> e = dynts.parse('2*GOOG')
	>>> e
	2.0 * goog
	>>> len(e)
	2
	>>> list(e)
	[2.0, goog]
	>>> ts = dynts.evaluate(e).unwind()
	>>> ts
	TimeSeries:numpy:2.0 * goog
	>>> len(ts)
	251


Requirements
=====================
There are few requirements that must be met:

* python_ 2.6 up to python 3.3.
* numpy_ version 1.5.1 or higher for arrays and matrices.
* ply_ version 3.3 or higher, the building block of the DSL_.
* ccy_ for date and currency manipulation.

R backend
===============================
Depending on the back-end used, additional dependencies need to be met.
For example, there are back-ends depending on the following R packages:

* rpy2_ if an R_ TimeSeries back-end is used (default).
* zoo_ and PerformanceAnlytics_ for the ``zoo`` back-end (currently the default one)
* timeSeries_ for the ``rmetrics`` back-end 

Installing rpy2_ on Linux is straightforward, on windows it requires the
`python for windows`__ extension library.

Optional Requirements
===============================

* cython_ for performance. The library is not strictly dependent on cython, however its usage
  is highly recommended. If available several python modules will be replaced by more efficient compiled C code.
* xlwt_ to create spreadsheet from timeseries.
* matplotlib_ for plotting.
* djpcms_ for the ``web.views`` module.

__ http://sourceforge.net/projects/pywin32/files/


.. _running-tests:

Running Tests
=================
There are three types of tests available:

* ``regression`` for unit and regression tests.
* ``profile`` for analysing performance of different backends and impact of cython_.
* ``bench`` same as ``profile`` but geared towards speed rather than profiling.
  
From the distribution directory type::
	
	python runtests.py
	
This will run by default the regression tests. To run a profile test
type::

	python runtests.py -t profile <test-name>
	
where ``<test-name>`` is the name of a profile test.
To obtain a list of available tests for each test type, run::

	python runtests.py --list

for regression, or:: 

	python runtests.py -t profile --list
	
for profile, or::

	python runtests.py -t bench --list
	
from benchmarks.
	
If you access the internet behind a proxy server, pass the ``-p`` option, for example::

	python runtests.py -p http://myproxy.com:80

It is needed since during tests some data is fetched from google finance.

To access coverage of tests you need to install the coverage_ package and run the tests using::

	coverage run runtests.py
	
and to check out the coverage report::

	coverage report -m


JavaScript
------------
* jQuery flot_ plugin for web graphs.
* jQuery Sparklines_ plugin for inline plotting.



Community
=================
Trying to use an IRC channel **#dynts** on ``irc.freenode.net``
(you can use the webchat at http://webchat.freenode.net/).

If you find a bug or would like to request a feature, please `submit an issue`__.

__ http://github.com/quantmind/dynts/issues
    
.. _numpy: http://numpy.scipy.org/
.. _ply: http://www.dabeaz.com/ply/
.. _rpy2: http://rpy.sourceforge.net/rpy2.html
.. _DSL: http://en.wikipedia.org/wiki/Domain-specific_language
.. _R: http://www.r-project.org/
.. _ccy: http://code.google.com/p/ccy/
.. _zoo: http://cran.r-project.org/web/packages/zoo/index.html
.. _PerformanceAnlytics: http://cran.r-project.org/web/packages/PerformanceAnalytics/index.html
.. _timeSeries: http://cran.r-project.org/web/packages/timeSeries/index.html
.. _Python: http://www.python.org/
.. _xlwt: http://pypi.python.org/pypi/xlwt
.. _matplotlib: http://matplotlib.sourceforge.net/
.. _djpcms: http://djpcms.com
.. _coverage: http://nedbatchelder.com/code/coverage/
.. _cython: http://www.cython.org/
.. _flot: http://code.google.com/p/flot/
.. _Sparklines: http://www.omnipotent.net/jquery.sparkline/

CHANGELOG.rst
Development
================================
* Web stuff moved under the :mod:`web` module.
* Improved documentation.
* Several bug fixes.
* Introduced :attr:`dynts.conf.Settings.desc` for specifying the order of dates in timeseries.
* Added `about` menu to :ref:`Ecoplot plugin <ecoplot-web>`.
* :ref:`Ecoplot plugin <ecoplot-web>` updated to work with jQuery_ 1.6.1. Changes are mainly in the
  way checked elements are handled::
  
    elem.prop({'checked':true});
    elem.is(':checked')
    
* **163 Tests**, **70% test coverage**

Ver. 0.4.1 - 2011 May 19
================================
* First Python 3 compatible version!
* numpy_ set as default timeseries backend.
* Removed dependency on rpy2_.
* Upgraded javascript plugin to work with flot_ 0.7 and djpcms_ 0.9.
* All timeseries functions must have a documentation, otherwise ``docs`` tests wont pass.
* **118 Tests**, **70% test coverage**

Ver. 0.4.0 - 2011 Feb 15
================================
* Dropped python 2.5 compatibility and moved towards python 3 support. The idea is to
  have the library working on python 2.6, 2.7 and the 3 series by the next version.
* Regular expression for ``DSL`` IDs changed to::

    r'`[^`]*`|[a-zA-Z_][a-zA-Z_0-9:@]*'
     
* Moved ``tests`` module outside ``dynts`` package. It now contains three types of tests:
  
  * ``regression`` for unit and regression tests.
  * ``profile`` for analysing performance of different backends and impact of cython_.
  * ``bench`` same as ``profile`` but geared towards speed rather than profiling.

  Check :ref:`Running tests <running-tests>` for more information.  	
* Added fallback tests and split ``zoo`` and ``numpy`` timeseries backend tests.
* Added sphinx_ extension in :mod:`dynts.web.dyntsx` for displaying the list of timeseries functions available.
* Created the ``lib`` module where cython_ extensions will be placed. Cython extensions will be compiled only
  if cython is available, otherwise a fall-back pure python library will be used.
* Added simple statistics pivot table.
* Bug fixes.
* **118 Tests**, **70% test coverage**

Ver. 0.3.3 - 2010 Nov 16
===================================
* Bug fix in :func:`dynts.TimeSeries.series` generator.
* **73 Tests**

Ver. 0.3.2 - 2010 Oct 20
======================================
* Development status set to ``Beta``.
* Added ``test`` module which implements shortcut functions for testing.
* Support for binary operations ``+-*\`` in place.
* Added ``series_info`` parameter to :ref:`flot formatter <formatters-flot>`.
* :ref:`Ecoplot plugin <ecoplot-web>` handles series options from server.
* More documentation.
* **10** ``dsl`` functions.
* **72 tests**. 

Ver. 0.3.1 - 2010 Oct 10
=================================
* Removed all dependencies from ``setup`` module.
* Added new :mod:`dynts.maths` module for mathematics and statistics.
* Passing ``request`` to :func:`dynts.web.views.TimeSeriesView.getdata`.
* Bug fix in :mod:`dynts.dsl` which was crashing the parser when using symbol names starting with a numeric value.
* **5** ``dsl`` functions.
* **61 tests**. 

Ver. 0.3.0 - 2010 Oct 06
==================================
* Added initial support for XY series in :class:`dynts.xydata`.
* :class:`dynts.TimeSeries` derived from :class:`dynts.DynData`.
* Added the :attr:`dynts.DynData.info` attribute for storing additional information about data. 
* Added a simple ``scatter`` function for performing scatter plots.
* Added logging parameter in :func:`dynts.evaluate`.
* Formatters are instances rather than functions.
* Can specify backend in :func:`dynts.evaluate`.
* Introduced :func:`dynts.tsname` for creating names for a mutivariate timeseries.
* Introduced :func:`dynts.merge` for merging two or more :class:`dynts.TimeSeries`.
* Refactored jQuery plugin ``ecoplot.js``.
* **5** ``dsl`` functions.
* **60 tests**. 

Ver. 0.2.0 - 2010 Sep 24
====================================
* Development moved to github http://github.com/quantmind/dynts
* Added skiplist python implementation.
* Added ccy_ to dependencies.
* **4** ``dsl`` functions.
* **53 tests**.

Ver. 0.1.0  - 2010 Sep 12
====================================
* First release to PyPi in pre-alpha.


.. _numpy: http://numpy.scipy.org/
.. _cython: http://www.cython.org/
.. _ccy: http://code.google.com/p/ccy/
.. _djpcms: http://djpcms.com/
.. _sphinx: http://sphinx.pocoo.org/
.. _flot: http://code.google.com/p/flot/
.. _rpy2: http://rpy.sourceforge.net/rpy2.html
.. _jQuery: http://jquery.com/

dynts/web/flot.py
from datetime import datetime, date

from dynts.utils.py2py3 import is_string
from dynts.utils.anyjson import JSONdatainfo, JSONobject

EPOCH = 1970
_EPOCH_ORD = date(EPOCH, 1, 1).toordinal()


def pydate2flot(dte):
    year, month, day, hour, minute, second = dte.timetuple()[:6]
    days = date(year, month, 1).toordinal() - _EPOCH_ORD + day - 1
    hours = days*24 + hour
    minutes = hours*60 + minute
    seconds = minutes*60 + second
    if isinstance(dte,datetime):
        return 1000*seconds + 0.001*dte.microsecond
    else:
        return 1000*seconds
    

class MultiPlot(JSONdatainfo):
    '''Class holding several :class:`dynts.web.flot.Flot` instances.
    
    .. attribute:: plots
    
        list of :class:`dynts.web.flot.Flot` instances.
        
    .. attribute:: info
    
        Additional serialisable data
    '''
    def __init__(self, flot = None, info = None):
        super(MultiPlot,self).__init__(data = [], info = info)
        self.add(flot)
        
    def add(self, flot):
        '''Add a new :class:`dynts.web.flot.Flot` instance to :attr:`dynts.web.flot.MultiPlot.plots`.'''
        if isinstance(flot,Flot):
            self.data.append(flot)
        
    def todict(self):
        return {'type': 'multiplot',
                'info': self.info,
                'plots': [plot.todict() for plot in self.data]}
    

class Flot(JSONobject):
    '''A single plot which can be a timeseries or a XY plot.'''
    allowed_types = ['xy','timeseries','scatter']
    def __init__(self, name = '', type = None, shadowSize = None, **kwargs):
        if type not in self.allowed_types:
            type = 'xy'
        self.name   = name
        self.type   = type
        self.series = []
        df = {}
        self.options = df
        if shadowSize:
            df['shadowSize'] = shadowSize
        
    def add(self, serie):
        if isinstance(serie,Serie):
            self.series.append(serie)
        
    def todict(self):
        od = super(Flot,self).todict()
        od['series'] = [serie.todict() for serie in self.series]
        return od


class Serie(JSONobject):
    
    def __init__(self, label = '', data = None,
                 color = None, shadowSize = None, yaxis = 1, xaxis = 1,
                 extratype = None, **kwargs):
        self.label = label
        if data is None:
            data = ()
        if extratype == 'date':
            ndata = []
            for a,b,dt in data:
                ndata.append((a,b,pydate2flot(dt)))
            data = ndata
        self.data = data
        for k,v in kwargs.items():
            setattr(self,k,v)
        self.extratype = extratype
        if is_string(color):
            if not color.startswith('#'):
                color = '#%s' % color
        self.xaxis = xaxis
        self.yaxis = yaxis
        if color:
            self.color = color
        if shadowSize:
            self.shadowSize = shadowSize
            
    def __str__(self):
        return self.label
    __repr__ = __str__


